src/ckb/channel.rs

```diff
@@ -3071,39 +3228,71 @@ impl ChannelActorState {
+    // `apply` = false means we want to mock the shutdown transaction and only for calculating the fee
+    pub fn build_shutdown_tx(
+        &self,
+        apply: bool,
+    ) -> Result<(TransactionView, [u8; 32]), ProcessingChannelError> {
         let (
             local_shutdown_script,
             remote_shutdown_script,
             local_shutdown_fee,
             remote_shutdown_fee,
-        ) = match (
-            self.local_shutdown_script.clone(),
-            self.remote_shutdown_script.clone(),
-            self.local_shutdown_fee,
-            self.remote_shutdown_fee,
-        ) {
-            (
-                Some(local_shutdown_script),
-                Some(remote_shutdown_script),
-                Some(local_shutdown_fee),
-                Some(remote_shutdown_fee),
-            ) => (
+        ) = if apply {
+            // Don't use get_local_shutdown_script and get_remote_shutdown_script here
+            // as they will panic if the scripts are not present.
+            // This function may be called in a state where these scripts are not present.
+            let (
                 local_shutdown_script,
                 remote_shutdown_script,
-                local_shutdown_fee,
-                remote_shutdown_fee,
-            ),
-            _ => {
-                return Err(ProcessingChannelError::InvalidState(format!(
-                    "Shutdown scripts are not present: local_shutdown_script {:?}, remote_shutdown_script {:?}, local_shutdown_fee {:?}, remote_shutdown_fee {:?}",
+                local_shutdown_fee_rate,
+                remote_shutdown_fee_rate,
+            ) = match (
+                self.local_shutdown_script.clone(),
+                self.remote_shutdown_script.clone(),
+                self.local_shutdown_fee_rate,
+                self.remote_shutdown_fee_rate,
+            ) {
+                (
+                    Some(local_shutdown_script),
+                    Some(remote_shutdown_script),
+                    Some(local_shutdown_fee_rate),
+                    Some(remote_shutdown_fee_rate),
+                ) => (
+                    local_shutdown_script,
+                    remote_shutdown_script,
+                    local_shutdown_fee_rate,
+                    remote_shutdown_fee_rate,
+                ),
+                _ => {
+                    return Err(ProcessingChannelError::InvalidState(format!(
+                    "Shutdown scripts are not present: local_shutdown_script {:?}, remote_shutdown_script {:?}, local_shutdown_fee_rate {:?}, remote_shutdown_fee_rate {:?}",
                     &self.local_shutdown_script, &self.remote_shutdown_script,
-                    &self.local_shutdown_fee, &self.remote_shutdown_fee
+                    &self.local_shutdown_fee_rate, &self.remote_shutdown_fee_rate
                 )));
-            }
+                }
+            };
+
+            let tx_size = self
+                .build_shutdown_tx(false)
```

Comment:
Yes, I considered that `build_shutdown_tx` may not build a valid shutdown transaction, when we're calculating the transaction size, we always use two dummy scripts to avoid this issue.

I'm also hesitant to reuse this function, one obvious benefit is reusing code, another advantage is we can always keep up to date if the logic for build shutdown transaction changed. But at the same time, it maybe cause some confusion, especially when `build_shutdown_tx(true)` will invoke `build_shutdown_tx(false)`.

I will try to add a new function to achieve this.